Fix #2681: Module Splitting Support

gzm0 opened the pull request.

gzm0 commented:
At the moment, the overall diff is identical to https://github.com/scala-js/scala-js/pull/4111

gzm0 commented:
Benchmarks for the single module case (i.e. default config).

![speed](https://user-images.githubusercontent.com/1140323/93361176-dc8e4f80-f844-11ea-8add-dfe5676d5e40.png)

This shows us that when module splitting is disabled, we do not regress on linking speed after this PR.

sjrd commented:
I'll probably have to go over this bits by bits ^^

By any chance, would you have some advice on in which order to read things in the big commit (Add Module Splitting support in the linker)?

gzm0 commented:
> By any chance, would you have some advice on in which order to read things in the big commit (Add Module Splitting support in the linker)?

1. Review `ModuleSet`.
2. Observe (informally, look at code if you need to convince yourself):
    - The top level export trees and module initializers have a (public) `moduleID` parameter.
    - The linker writes to a directory instead of a single file.
3. Review `LinkerFrontendImpl`
4. Review `ModuleSplitter` & `ModuleAnalyzer`.
5. Review the `ModuleAnalyzer` implementations.
6. Review `ModuleContext`, `Emitter.ClassID`, `Emitter`.
7. Review `VarGen`.
8. Review the rest of `emitter` (boring, mostly more `WithGlobal` and more implicits).
9. Review `OutputWriter`.
10. At this point, you understand how the splitting works. Review everything else (in whatever order you want).

sjrd commented:
Thank you! That will be very helpful. :)

gzm0 commented:
I forgot about the `Analyzer`. I suggest you review it after the emitter.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:
```diff
@@ -127,6 +202,41 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
+    } else {
+      val export = config.moduleKind match {
+        case ModuleKind.NoModule =>
+          throw new AssertionError("non-leaf module in NoModule mode")
```
```suggestion
          throw new AssertionError("non-public module in NoModule mode")
```
?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:
```diff
@@ -41,50 +40,120 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
+      val setterIdent = globalVarIdent(setterField, scope, origName)
+      val x = Ident("x")
+      val setter = FunctionDef(setterIdent, List(ParamDef(x, rest = false)), {
+          Assign(VarRef(ident), VarRef(x))
```
```suggestion
        Assign(VarRef(ident), VarRef(x))
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:
```diff
@@ -41,50 +40,120 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
+    val varDef = genLet(ident, mutable = true, value)
+
+    if (config.moduleKind == ModuleKind.ESModule && !moduleContext.public) {
+      val setterIdent = globalVarIdent(setterField, scope, origName)
```
I don't think it makes sense to give the original name of the field to the setter. They are separate entities.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/KnowledgeGuardian.scala:
```diff
@@ -39,17 +40,26 @@ private[emitter] final class KnowledgeGuardian(config: Emitter.Config) {
+    val staticFieldMirrors = computeStaticFieldMirrors(moduleSet)
 
-    var objectClass: LinkedClass = null
+    // Object is optional, because the module splitter might remove everything.
+    var objectClass: Option[LinkedClass] = None
```
Have you considered short-cutting this situation right at the start of `Emitter.emit()`, before calling `emitInternal()`? That way we wouldn't have to deal with this corner case everywhere else in the emitter.

gzm0 replied:
Hmmm... I can, but this is the only place that needs changing IIUC. So I'm not sure short-cutting at the beginning is simpler (or clearer). I do not feel strongly about it though. So if you prefer that, I can do that.

sjrd replied:
I don't really have a strong preference either. We can keep it as is.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/KnowledgeGuardian.scala:
```diff
@@ -126,14 +138,17 @@ private[emitter] final class KnowledgeGuardian(config: Emitter.Config) {
+      for {
+        module <- moduleSet.modules
+        export <- module.topLevelExports
+      } yield {
```
The `yield` shouldn't be there, I think. This for comprehension is in statement position, so its result is ignored.
```suggestion
      } {
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -631,8 +633,23 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
+          if (needToUseGloballyMutableVarSetter(scope)) {
+            unnest(rhs) { (rhs, env0) =>
+              implicit val env = env0
+              js.Apply(globalVar("u", scope), transformExprNoChar(rhs) :: Nil)
```
Couldn't the static field have a `char` type? I think you need to use `transformExpr` with an expected type:
```suggestion
              js.Apply(globalVar("u", scope), transformExpr(rhs, select.tpe) :: Nil)
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/OutputWriter.scala:
```diff
@@ -0,0 +1,110 @@
+
+    for {
+      currentFiles <- outputImpl.listFiles()
+      reports <- Future.traverse(moduleSet.modules)(m =>
+          ioThrottler.throttle(writeModule(m.id)))
+      _ <- Future.traverse(filesToRemove(currentFiles, reports))(f =>
+          ioThrottler.throttle(outputImpl.delete(f)))
```
```suggestion
      reports <- Future.traverse(moduleSet.modules)(
          m => ioThrottler.throttle(writeModule(m.id)))
      _ <- Future.traverse(filesToRemove(currentFiles, reports))(
          f => ioThrottler.throttle(outputImpl.delete(f)))
```
or
```suggestion
      reports <- Future.traverse(moduleSet.modules) { m =>
        ioThrottler.throttle(writeModule(m.id))
      }
      _ <- Future.traverse(filesToRemove(currentFiles, reports)) { f =>
        ioThrottler.throttle(outputImpl.delete(f))
      }
```

sjrd reviewed (changes requested):
I reviewed the steps 1 to 9 so far (include 6b for the `Analyzer`). Overall this looks really excellent! I have very minor comments here and there, but nothing major.

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Linker.scala:
```diff
@@ -12,19 +12,197 @@
+   */
+  private val fileFieldRe = """(?m)([,{])\s*"file"\s*:\s*".*"\s*([,}])""".r
+
+  /** Patches the JS file content to have the provided source map link (or none) */
+  private def patchJSFileContent(content: String,
+      sourceMapURI: Option[URI]): String = {
```
Could you elaborate a bit on why those patches are necessary, and what they do exactly?

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Report.scala:
```diff
@@ -0,0 +1,155 @@
+   *  An incremental linker must return all public modules here, even if some
+   *  were not re-written during this linking run.
+   */
+  def publicModules: Iterable[Report.Module]
```
We should have an known-to-be-immutable data type, here. I think it should be `immutable.Seq[Report.Module]`, if not `List[Report.Module]`.

gzm0 replied:
I'd rather not commit to the ordering of the modules. Is `immutable.Iterable` OK?

sjrd replied:
Yes, `immutable.Iterable` works too.

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Report.scala:
```diff
@@ -0,0 +1,155 @@
+import org.scalajs.linker.interface.unstable.ReportImpl
+
+/** Information about a linker run. */
+abstract class Report private[interface] {
```
When we have a visibility modifier for the constructor, we always write the `()`:
```suggestion
abstract class Report private[interface] () {
```

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Report.scala:
```diff
@@ -0,0 +1,155 @@
+
+object Report {
+  /** Information about a module produced by the linker. */
+  abstract class Module private[interface] {
```
```suggestion
  abstract class Module private[interface] () {
```

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Report.scala:
```diff
@@ -0,0 +1,155 @@
+
+  /** Serializes this [[Report]] to a byte array.
+   *
+   *  A report serialized with the exact same linker version is guaranteed to
```
```suggestion
   *  A report serialized with the exact same linker version is guaranteed to be
```

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Report.scala:
```diff
@@ -0,0 +1,155 @@
+
+    private def writeModuleKind(kind: ModuleKind): Unit = {
+      val i = kind match {
+        case ModuleKind.NoModule => 0
+        case ModuleKind.ESModule => 1
+        case ModuleKind.CommonJSModule => 2
```
```suggestion
        case ModuleKind.NoModule       => 0
        case ModuleKind.ESModule       => 1
        case ModuleKind.CommonJSModule => 2
```

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/unstable/OutputDirectoryImpl.scala:
```diff
@@ -0,0 +1,42 @@
+   *
+   *  Writing should only result in a file write if the contents of the file
+   *  actually changed. Further, if the underlying filesystem allows it, the
+   * file should be written atomically.
```
```suggestion
   *  file should be written atomically.
```

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/unstable/OutputDirectoryImpl.scala:
```diff
@@ -0,0 +1,42 @@
+      implicit ec: ExecutionContext): Future[Unit]
+
+  /** Lists all the files in the directory. */
+  def listFiles()(implicit ec: ExecutionContext): Future[Iterable[String]]
```
This is less important than in `Report`, but I would prefer an immutable data here as well, since it is an asynchronous result.

sjrd commented on linker/js/src/main/scala/org/scalajs/linker/NodeOutputDirectory.scala:
```diff
@@ -0,0 +1,50 @@
+        else ByteBuffer.allocateDirect(buf.remaining()).put(buf).typedArray()
+
+      cbFuture[Unit](NodeFS.writeFile(path, data, _))
+        .map(_ => buf.position(buf.limit()))
```
This is a bit misleading. If `buf` does not have a TypedArray, it will already have been advanced to its `limit()` by `put(buf)`. I suggest doing that synchronously in the "then" branch of the `if (buf.hasTypedArray())` to make both branches behave the same.

sjrd commented on linker/jvm/src/main/scala/org/scalajs/linker/PathOutputDirectory.scala:
```diff
@@ -0,0 +1,204 @@
+      blocking {
+        val builder = Iterable.newBuilder[String]
+
+        /* Files.list(...) would be simpler, but it requires Java Streams which
+         * are, at the time of this writing, not implemented in Scala.js.
```
This code is in `jvm/`, so it could use Streams if that makes it easier.

sjrd commented on linker/jvm/src/main/scala/org/scalajs/linker/PathOutputDirectory.scala:
```diff
@@ -0,0 +1,204 @@
+        chan.read(readBuf, pos, null, Handler)
+      }
+
+      object Handler extends CompletionHandler[Integer, Null] {
```
```suggestion
      object Handler extends CompletionHandler[Integer, Unit] {
```
?

sjrd reviewed (commented):
Some more review until `linker/jvm/src/test/scala/org/scalajs/linker/PathOutputDirectoryTest.scala`, excluded.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala:
```diff
@@ -74,12 +75,14 @@ final class Emitter(config: Emitter.Config) {
     moduleKind match {
       case ModuleKind.NoModule =>
-        val topLevelVars = unit.topLevelExports.map(_.exportName)
+        val topLevelVars = moduleSet.modules
```
No, we cannot. We can have 0 or 1 module here. Not sure how to improve this though.

sjrd replied:
Perhaps just `assert(moduleSet.modules.size <= 1)` and then get `moduleSet.modules.head`?

It may be less "type-safe" but it would be clear what's happening.

gzm0 commented:
I have addressed the first round of comments. All questions I have not answered in the review thread should be answered in code comments.

sjrd commented on linker/shared/src/test/scala/org/scalajs/linker/LinkerTest.scala:
```diff
@@ -84,4 +116,41 @@ class LinkerTest {
 
+  @Test
+  def testLegacyAPI(): AsyncResult = await {
+    val classDefs = Seq(
+        mainTestClassDef({
+          consoleLog(StringLiteral("Hello world!"))
+        })
+    )
```
`helloWorldClassDefs`?

sjrd reviewed (commented):
And that's the end of the review for the big Linker commit.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/MemOutputDirectory.scala:
```diff
@@ -0,0 +1,60 @@
+/** OutputDirectory that simply writes to memory. */
+sealed trait MemOutputDirectory extends OutputDirectory {
+  /** Content of the file with `name` or `None` if the file was not written. */
+  def content(name: String): Option[Array[Byte]]
```
By using the file names in the `Report`. I'm not opposed to exposing a way of listing files (or even immutable access to the underlying `Map`), but I was trying to adhere to "expose the minimum required ".

sjrd replied:
> By using the file names in the `Report`.

Ah yes, that makes sense. Let's keep it minimal, then.

gzm0 commented:
I forgot to mention: I also rebased on `master`. There was a trivial conflict in `Build.scala` (adjacent, independent lines modified).

sjrd commented on project/Build.scala:
```diff
@@ -1732,47 +1749,60 @@ object Build {
-      jsEnvInput in Test ++=
-        (setModuleLoopbackScript in Test).value.toList.map(Input.ESModule(_)),
+        val setDict = {
+          val dict = report.publicModules.map(m =>
+              s"${m.moduleID}: ${m.moduleID}"
+          ).mkString("{", ",", "}")
```
```suggestion
          val dict = report.publicModules.map(
              m => s"${m.moduleID}: ${m.moduleID}"
          ).mkString("{", ",", "}")
```
or even
```suggestion
          val dict = report.publicModules
            .map(m => s"${m.moduleID}: ${m.moduleID}")
            .mkString("{", ",", "}")
```

sjrd commented on project/Build.scala:
```diff
@@ -1732,47 +1749,60 @@ object Build {
+             * unlikely, though, given all the message passing for the com
+             * and all that.
+             */
+            val imports = report.publicModules.map(m =>
+                s"""import * as ${m.moduleID} from "./${escapeJS(m.jsFileName)}";\n"""
+            ).mkString
```
```suggestion
            val imports = report.publicModules
              .map(m => s"""import * as ${m.moduleID} from "./${escapeJS(m.jsFileName)}";\n""")
              .mkString
```

sjrd commented on project/Build.scala:
```diff
@@ -1732,47 +1749,60 @@ object Build {
+            List(Input.ESModule(outputFile.toPath))
+
+          case ModuleKind.CommonJSModule =>
+            val requires = report.publicModules.map(m =>
+                s"""var ${m.moduleID} = require("./${escapeJS(m.jsFileName)}");\n"""
+            ).mkString
```
```suggestion
            val requires = report.publicModules
              .map(m => s"""var ${m.moduleID} = require("./${escapeJS(m.jsFileName)}");\n""")
              .mkString
```

sjrd reviewed (changes requested):
Reviewed until `ScalaJSPluginInternal.scala` (excluded) in the sbt plugin commit.

gzm0 commented on Jenkinsfile:
```diff
@@ -136,7 +152,14 @@ def Tasks = [
     sbtretry ++$scala testSuiteEx$v/test &&
     sbtretry 'set scalaJSStage in Global := FullOptStage' \
         ++$scala testSuiteEx$v/test &&
-    sbtretry ++$scala testSuite$v/test:doc library$v/test compiler$v/test reversi$v/fastOptJS reversi$v/fullOptJS &&
```
There is the `settings/legacy-link-tasks` scripted test. Do you also want an invocation in the Jenkinsfile?

sjrd replied:
Ah no, that's probably good enough. I didn't get that far yet ^^

gzm0 commented:
@sjrd There is a subtle bug I have introduced when I moved `symbolRequirement` handling to the `Analyzer`. The main reason for this was that it's infeasible to handle optional requirements in the module analyzer (because it needs to make decisions based on method existence). Now in the `Analyzer`, `symbolRequirements` also contain module initializers. As such, `jl.Object` ends up statically depending on module entry points. This is of course total nonsense.

I will attempt to fix the problem by making module initializers first class citizens in the Analyzer. This is unfortunate, but I'm not sure I see another option.

gzm0 commented:
I have added a commit fixing the bug and adding a regression test.

sjrd commented on linker/shared/src/test/scala/org/scalajs/linker/MinModuleSplittingTest.scala:
```diff
@@ -0,0 +1,84 @@
+class MinModuleSplittingTest {
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  /** Smoke test to ensure modules do note get merged too much. */
```
```suggestion
  /** Smoke test to ensure modules do not get merged too much. */
```

sjrd commented on test-suite-linker/src/main/scala/org/scalajs/bootstrap/TestSuiteLinker.scala:
```diff
@@ -32,25 +32,27 @@ object QuickLinker {
+  @JSImport("fs", "promises")
   @js.native
-  private def basename(str: String): String = js.native
+  object PromisesFS extends js.Object {
```
```suggestion
  private object PromisesFS extends js.Object {
```

sjrd commented on test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/ExportsTest.scala:
```diff
@@ -1478,6 +1472,19 @@ class ExportsTest {
     }
   }
 
+  @Test def top_level_export_field_is_writiable_accross_modules(): Unit = {
```
```suggestion
  @Test def top_level_export_field_is_writable_accross_modules(): Unit = {
```

sjrd commented on sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:
```diff
@@ -122,6 +131,8 @@ private[sbtplugin] object ScalaJSPluginInternal {
         box.ensure(linkerImpl.clearableLinker(config))
       },
 
+      scalaJSLinker in legacyKey := (scalaJSLinker in key).value,
```
This is not right. It will silently break compatibility in ways that are hard to debug. Although it's a bit of a specialized use case, if a user had specified
```scala
scalaJSLinker in (Compile, fastOptJS) := something
```
that setting will now be ignored. IMO we should do the reverse. Set `scalaJSLinker in legacyKey` first, then derive `scalaJSLinker in key := (scalaJSLinker in legacyKey).value`.

gzm0 replied:
Writing to `scalaJSLinker` is specified to be an unstable API, but fair enough, for consistency with everything else, we might as well do it.

sjrd commented on sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:
```diff
@@ -141,6 +152,8 @@ private[sbtplugin] object ScalaJSPluginInternal {
         Tags.Tag(s"uses-scalajs-linker-$projectPart-$configPart-$stagePart")
       },
 
+      usesScalaJSLinkerTag in legacyKey := (usesScalaJSLinkerTag in key).value,
```
This one is fine because it is essentially a read-only key from the users' point of view.

sjrd commented on sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:
```diff
@@ -151,6 +164,8 @@ private[sbtplugin] object ScalaJSPluginInternal {
       scalaJSLinkerConfigFingerprint in key :=
         StandardConfig.fingerprint((scalaJSLinkerConfig in key).value),
 
+      moduleName in legacyKey := (moduleName in key).value,
```
This one is arguably OK because it will still affect the `artifactPath in legacyKey`. Nevertheless, I think we should also derive the new one from the old one, for consistency.

sjrd commented on sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala:
```diff
@@ -335,14 +410,21 @@ private[sbtplugin] object ScalaJSPluginInternal {
             ((moduleName in fullOptJS).value + "-opt.js")),
 
-      scalaJSLinkerConfig in fullOptJS ~= { prevConfig =>
+      scalaJSLinkerConfig in linkJSProd ~= { prevConfig =>
```
This is the most annoying one, as it is a very real use case for people to tweak `scalaJSLinkerConfig in (Compile, fullOptJS)`. We absolutely must not throw away their configuration, *especially* since it also affects the result of the legacy keys themselves.

So we should still have here `in fullOptJS`, and after that add
```scala
scalaJSLinkerConfig in linkJSDev := (scalaJSLinkerConfig in fastOptJS).value,
scalaJSLinkerConfig in linkJSProd := (scalaJSLinkerConfig in fullOptJS).value,
```

sjrd reviewed (changes requested):
All right, here is the end of the review, including the latest commit that fixes the module initializer issue.

There is one serious issue in `ScalaJSPluginInternal` (there are a few comments that relate to the same thing), but otherwise it looks very good.

gzm0 commented:
I have added a commit addressing the comments. (I have changed everything to be in terms of the `legacyKey` for consistency).

Regarding the names of the new tasks. I like your suggestion from #4111:

> Perhaps fastLinkJS and fullLinkJS? I'll keep thinking.

It keeps `fast` / `full` but introduces the name `link` which makes sense anyways. Probably better than introducing `dev` / `prod`.

For me, there are two open questions on this PR:
- Shall we rename `linkJSDev`  / `linkJSProd` (`linkJSFast` / `linkJSFull` might also be options, also in view of autocomplete: `fast<tab>` would give the old tasks. then again, visibility will probably handle that).
- Are there enough tests for the legacy integration (notably the one in `Linker`)?

sjrd commented:
> * Shall we rename `linkJSDev`  / `linkJSProd` (`linkJSFast` / `linkJSFull` might also be options, also in view of autocomplete: `fast<tab>` would give the old tasks. then again, visibility will probably handle that).

I would say that I prefer both `linkJSFast` and `fastLinkJS` over `linkJSDev`. I can't really make up my about which is better between `linkJSFast` and `fastLinkJS`. The latter *sounds* (literally) better to me, but that is very subjective.

> * Are there enough tests for the legacy integration (notably the one in `Linker`)?

For the programmatic API of the `Linker`, I think that's enough, yes. For the legacy sbt API, I would suggest adding a few more customizations in `legacy-link-tasks` that would exercise the other keys you fixed in the last commit (`scalaJSLinkerConfig in (Compile, fullOptJS)` notably).

gzm0 commented:
> The latter sounds (literally) better to me, but that is very subjective.

I'd tend to agree in fact. It speaks easier.

OTOH, `linkJSFast` is closer to how sbt tasks like `testOnly`, `testQuick` or `runMain` are named.

From the 6 possible permutations of `js`, `link` and `fast`, I think the following are acceptable:

- `linkJSFast`
- `fastLinkJS`
- `jsLinkFast`
- `jsFastLink`

`linkFastJS` and `fastJSLink` are bad, because they'd imply the JS is fast.

Personally, I think I prefer `linkJSFast` because of how it aligns with sbt's tasks and it goes from most to least important property of the task left to right (that argument might also work in favor of `jsLinkFast` though).

sjrd commented:
On the other hand, `fastLinkJS` is more similar with today's `fastOptJS` ...

Ah, naming in computer science!

gzm0 commented:
Let's settle on `fastLinkJS` then. My plan is to squash the commits and then perform the rename. Then re-push, including the commit-by-commit CI. Is that OK with you?

sjrd commented:
Sounds good. :)

gzm0 commented:
Rebased. Testing marathon is starting.

gzm0 commented:
@sjrd testing is reaching completion. Do you have further comments on this?

sjrd reviewed (approved):
Let's ship it! 🚀 ✨
